llama-index-integrations/graph_stores/llama-index-graph-stores-nebula/llama_index/graph_stores/nebula/nebula_property_graph.py

```diff
@@ -0,0 +1,932 @@
+CREATE TAG INDEX IF NOT EXISTS idx_Node__ ON `Node__`(`label`);
+CREATE EDGE INDEX IF NOT EXISTS idx_Relation__ ON `Relation__`(`label`);
+
+CREATE EDGE INDEX IF NOT EXISTS idx_meta__node_label__ ON `__meta__node_label__`(`label`);
+CREATE EDGE INDEX IF NOT EXISTS idx_meta__rel_label__ ON `__meta__rel_label__`(`label`);
+"""
+
+# Hard coded default schema, which is union of
+# document metadata: `file_path` STRING, `file_name` STRING, `file_type` STRING, `file_size` INT, `creation_date` STRING, `last_modified_date` STRING
+# llamaindex_node: `_node_content` STRING, `_node_type` STRING, `document_id` STRING, `doc_id` STRING, `ref_doc_id` STRING
+# introduced by PropertyGraph: `triplet_source_id` STRING
+DEFAULT_PROPS_SCHEMA = "`file_path` STRING, `file_name` STRING, `file_type` STRING, `file_size` INT, `creation_date` STRING, `last_modified_date` STRING, `_node_content` STRING, `_node_type` STRING, `document_id` STRING, `doc_id` STRING, `ref_doc_id` STRING, `triplet_source_id` STRING"
+
+
+class NebulaPropertyGraphStore(PropertyGraphStore):
+    """
+    NebulaGraph Property Graph Store.
+
+    This class implements a NebulaGraph property graph store.
+
+    You could go with NebulaGraph-lite freely on Google Colab.
+    - https://github.com/nebula-contrib/nebulagraph-lite
+    Or Install with Docker Extension(search in the Docker Extension marketplace) on your local machine.
+
+    Examples:
+        `pip install llama-index-graph-stores-nebula`
+        `pip install jupyter-nebulagraph`
+
+        Create a new NebulaGraph Space with Basic Schema:
+
+        ```jupyter
+        %load_ext ngql
+        %ngql --address 127.0.0.1 --port 9669 --user root --password nebula
+        %ngql CREATE SPACE IF NOT EXISTS llamaindex_nebula_property_graph(vid_type=FIXED_STRING(256));
+        ```
+    """
+
+    _space: str
+    _client: BaseExecutor
+    sanitize_query_output: bool
+    enhanced_schema: bool
+
+    def __init__(
+        self,
+        space: str,
+        client: Optional[BaseExecutor] = None,
+        username: str = "root",
+        password: str = "nebula",
+        url: str = "nebula://localhost:9669",
+        overwrite: bool = False,
+        props_schema: str = DEFAULT_PROPS_SCHEMA,
+        refresh_schema: bool = True,
+        sanitize_query_output: bool = False,  # We don't put Embedding-Like values as Properties
+        enhanced_schema: bool = False,
+    ) -> None:
+        self.sanitize_query_output = sanitize_query_output
+        self.enhcnaced_schema = enhanced_schema
+
+        self._space = space
+        if client:
+            self._client = client
+        else:
+            session_pool = SessionPool(
+                username,
+                password,
+                self._space,
+                [url_scheme_parse(url)],
+            )
+            session_pool.init()
+            self._client = session_pool
+        self._client.execute(DDL.render(props_schema=props_schema))
```

Comment:
we need to add mechanisms here to retry due to DDL creation is async and it takes time for all services in cluster to get synced(two heartbeats, by defaults it's 20s)